SW-2543 get the guided tour working again page 7 guided tour

JFSolution opened the pull request.

Commit:
fix guided tour not working
changed guided tour design
improve focus step
added guided tour as extra nav item

Commit:
fix translatable strings
add congratulations modal TODO in SW-2544

Commit:
adding cypress tests

Commit:
skip the docker not supported steps

Commit:
wrongly skipped both tests

JFSolution commented:
Cypress test run https://github.com/mrbeam/MrBeamPlugin/actions/runs/4224053530

khaledsherkawi commented on octoprint_mrbeam/static/css/hopscotch.min.css:
```diff
@@ -445,14 +445,15 @@ div.hopscotch-bubble .hopscotch-bubble-close.hide-all {
   display: none;
 }
 div.hopscotch-bubble .hopscotch-bubble-number {
-  background: transparent url(../img/sprite-green.png) 0 0 no-repeat;
```
This is a library. We cannot modify it. Please fine another way to do that.

JFSolution replied:
This library is very old and will not get any updates as it dropped the development. Thats why I decided to do it here as it will be only till we will use a newer one.

khaledsherkawi replied:
I understand this but nevertheless, I would not take that approach and especially not by removing classes and modifying JS

khaledsherkawi commented on octoprint_mrbeam/static/js/lib/hopscotch.js:
```diff
@@ -2727,21 +2727,21 @@
             if (buttons.showPrev) {
                 __p +=
-                    '<button class="hopscotch-nav-button prev hopscotch-prev">' +
+                    '<button class="btn prev hopscotch-prev">' +
```
This is a library. We cannot modify it. Please find another way to do that.

khaledsherkawi commented on octoprint_mrbeam/static/js/lib/hopscotch.min.js:
```diff
@@ -1318,19 +1318,19 @@
                         g.showPrev &&
                             (e +=
-                                '<button class="hopscotch-nav-button prev hopscotch-prev">' +
+                                '<button class="btn prev hopscotch-prev">' +
```
This is a minified file of the library. We cannot modify it. Please find another way to do that.

Commit:
SW-2543
- revert hopsctoch lib changes
- moved the changes in our files with a observer and a css override
- moved backdrop observer in observer helper file

JFSolution commented:
https://github.com/mrbeam/MrBeamPlugin/actions/runs/4242033931

khaledsherkawi commented on octoprint_mrbeam/static/js/app/helpers/mutation-observer.js:
```diff
@@ -0,0 +1,119 @@
+            }
+            // Backdrop Temporary Solution - end
+
+            guidedTourOverride();
```
```suggestion
            // Workaround to override library styling upon dynamically rendering new elements
            guidedTourOverride();
```

khaledsherkawi commented on octoprint_mrbeam/static/js/app/helpers/mutation-observer.js:
```diff
@@ -0,0 +1,119 @@
+const mutationCallback = function (mutationsList, observer) {
+    for (let mutation of mutationsList) {
+        if (mutation.type === "childList") {
+            // Todo: should be removed once OctoPrint is updated
```
Optional; It would be nice to move this out into a separate function as well

khaledsherkawi commented on octoprint_mrbeam/templates/tab_workingarea.jinja2:
```diff
@@ -51,8 +51,7 @@
 													<div class="row-fluid print-control btn-group" style="display: none;" data-bind="visible: loginState.isUser">
 														<button id="laser_button" data-test="working-area-laser-button" class="btn btn-danger span6"
-                                                                data-bind="click: conversion.show_conversion_dialog, enable: {{ 'true' if isDevelop and 'enableLaserButton' in request.args else 'false'}} || (state.isOperational() && state.isReady() && !state.isPrinting() && !state.isPaused() && loginState.isUser() && !workingArea.working_area_empty()), attr: {title: state.titlePrintButton}"
-                                                                id="job_print">
```
What exactly is this ID? Were there 2 IDs for the same element?

JFSolution replied:
Yes there were two ids before thats why the whole guided tour didn't work before.
I seached for it and couldn'T find its usage, I came from the original octoprint print button.

Commit:
SW-2543 fix failing test

JFSolution committed:
Update octoprint_mrbeam/static/js/app/helpers/mutation-observer.js

Co-authored-by: khaledsherkawi <54568489+khaledsherkawi@users.noreply.github.com>

Commit:
SW-2543
- moved backdrop fix to function
- fixed test

Commit:
Merge branch 'feature/SW-2538/SW-2543-get-the-guided-tour-working-again-page-7-guided-tour' of github.com:mrbeam/MrBeamPlugin into feature/SW-2538/SW-2543-get-the-guided-tour-working-again-page-7-guided-tour

Commit:
SW-2543 revert changes to design library test
- select randome file for guided tour

JFSolution commented:
https://github.com/mrbeam/MrBeamPlugin/actions/runs/4251630912

Commit:
SW-2543 fix failing test

JFSolution committed:
Merge branch 'alpha' into feature/SW-2538/SW-2543-get-the-guided-tour-working-again-page-7-guided-tour

khaledsherkawi commented on octoprint_mrbeam/static/js/app/helpers/mutation-observer.js:
```diff
@@ -0,0 +1,118 @@
+            }
+        });
+
+        // only add this observer once
```
```suggestion
        // only add this observer once the user is going through the tour guide
```

khaledsherkawi commented on octoprint_mrbeam/static/js/app/view-models/tour.js:
```diff
@@ -512,7 +530,7 @@ $(function () {
                 nextBtn: gettext("Next"),
-                // prevBtn: "Back",
+                prevBtn: gettext("Back"),
                 // doneBtn: "Done",
                 // skipBtn: "Skip",
```
```suggestion
```

JFSolution committed:
Apply suggestions from code review

Co-authored-by: khaledsherkawi <54568489+khaledsherkawi@users.noreply.github.com>
